app/src/main/java/protect/card_locker/MainActivity.java

```diff
@@ -511,9 +495,68 @@ private void updateLoyaltyCardList(boolean updateCount) {
         }
     }
 
+    private void processBarcodeValues(BarcodeValues barcodeValues, String group) {
+        if (barcodeValues.isEmpty()) {
+            return;
+        }
+        Intent newIntent = new Intent(getApplicationContext(), LoyaltyCardEditActivity.class);
+        Bundle newBundle = new Bundle();
+        newBundle.putString(LoyaltyCardEditActivity.BUNDLE_BARCODETYPE, barcodeValues.format());
+        newBundle.putString(LoyaltyCardEditActivity.BUNDLE_CARDID, barcodeValues.content());
+        if (group != null) {
+            newBundle.putString(LoyaltyCardEditActivity.BUNDLE_ADDGROUP, group);
+        }
+        newIntent.putExtras(newBundle);
+        startActivity(newIntent);
+    }
+
+    private void onSharedIntent(Intent intent) {
+        String receivedAction = intent.getAction();
+        String receivedType = intent.getType();
+
+        // Check if an image was shared to us
+        if (Intent.ACTION_SEND.equals(receivedAction)) {
+            if (receivedType.startsWith("image/")) {
+                BarcodeValues barcodeValues;
+                try {
+                    Bitmap bitmap;
+                    try {
+                        Uri data = intent.getParcelableExtra(Intent.EXTRA_STREAM);
+                        bitmap = Utils.retrieveImageFromUri(this, data);
+                    } catch (IOException e) {
+                        Log.e(TAG, "Error getting data from image file");
+                        e.printStackTrace();
+                        Toast.makeText(this, R.string.errorReadingImage, Toast.LENGTH_LONG).show();
+                        finish();
+                        return;
+                    }
+
+                    barcodeValues = Utils.getBarcodeFromBitmap(bitmap);
+
+                    if (barcodeValues.isEmpty()) {
+                        Log.i(TAG, "No barcode found in image file");
+                        Toast.makeText(this, R.string.noBarcodeFound, Toast.LENGTH_LONG).show();
+                        finish();
+                        return;
+                    }
+                } catch (NullPointerException e) {
+                    Toast.makeText(this, R.string.errorReadingImage, Toast.LENGTH_LONG).show();
+                    finish();
+                    return;
+                }
+                Bundle inputBundle = intent.getExtras();
+                String group = inputBundle != null ? inputBundle.getString(LoyaltyCardEditActivity.BUNDLE_ADDGROUP) : null;
```

Comment:
These two lines seem useless. No gallery or file manager app would ever know what groups exist in Catima and which one to send to it. I think it's best to just leave these 2 lines out to keep the code more readable.

```suggestion
```